<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 12 Conclusions | Bioinformatics" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2024-12-09" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 12 Conclusions | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a href="sec-refresher.html" id="toc-sec-refresher"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html" id="toc-sec-vis"><span class="toc-section-number">2</span> Data visualisation</a>
<a href="sec-obj.html" id="toc-sec-obj"><span class="toc-section-number">3</span> High-level data structures</a>
<a href="sec-biostrings.html" id="toc-sec-biostrings"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a>
<a href="sec-norm.html" id="toc-sec-norm"><span class="toc-section-number">5</span> Data normalisation: centring, scaling, quantile normalisation</a>
<a href="sec-mlintro.html" id="toc-sec-mlintro"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a href="sec-testing.html" id="toc-sec-testing"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-dimred.html" id="toc-sec-dimred"><span class="toc-section-number">8</span> Unsupervised learning: dimensionality reduction</a>
<a href="sec-ul.html" id="toc-sec-ul"><span class="toc-section-number">9</span> Unsupervised learning: clustering</a>
<a href="sec-sl.html" id="toc-sec-sl"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html" id="toc-sec-biovis"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a id="active-page" href="sec-ccl.html" id="toc-sec-ccl"><span class="toc-section-number">12</span> Conclusions</a><ul class="toc-sections">
<li class="toc"><a href="#additional-exercise"> Additional exercise</a></li>
</ul>
<a href="sec-si.html" id="toc-sec-si"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-ccl" class="section level1" number="12">
<h1>
<span class="header-section-number">Chapter 12</span> Conclusions</h1>
<p>In this course, we have consolidated our skills in data analysis and
visualisation using R. In particular, this course has focused on
interpretation and understanding of the outputs.</p>
<p>We have also learned and applied new tools, in particular how to
manipulate sequence data using <code>Biostrings</code> and how to use statistical
learning tools to explore data and identify patterns of biological
interest. In terms of statistical testing and machine learning, it can
be useful to summarise the different classes of techniques we have
touch on using the figures below. The grid on each of these figures
represents a matrix with quantitative values with features (genes,
proteins, transcripts, …) along the rows and samples along the
columns. Annotations of features or samples are presented as coloured
boxes on the right or top if the matrices.</p>
<p>In hypothesis test, we start with quantitative data and an
experimental design, i.e. sample annotation that group samples in
biologically relevant groups. The output of hypothesis testing is a
set of metrics for each features (a p-value, an adjusted p-value, a
fold-change, …) that informs whether that feature shows any
difference between the biological groups of interest.</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-103"></span>
<p class="caption marginnote shownote">
Figure 12.1: Hypothesis testing.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-103-1.png" alt="Hypothesis testing." width="672">
</div>
<p>When performing clustering (unsupervised machine learning), we only have our
quantitative data as input, and the clustering algorithm (whether
k-means, hierarchical clustering among many others) suggests a set of
groups to cluster the features (as shown below) or samples.</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-104"></span>
<p class="caption marginnote shownote">
Figure 12.2: Clustering of features.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-104-1.png" alt="Clustering of features." width="672">
</div>
<p>When performing dimensionality reduction with, for example PCA, one
starts with a <em>n</em> by <em>m</em> data set as input to reduce the number of
dimensions in either direction. On the figure below, the number of
features <em>n</em> was reduced to 2 to, typically, visualise the sample
along a scatter plot.</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-105"></span>
<p class="caption marginnote shownote">
Figure 12.3: Dimensionality reduction.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-105-1.png" alt="Dimensionality reduction." width="672">
</div>
<p>In classification (supervised machine learning), we need labelled
data, i.e. a set of sample (top on the figures below) or features
(bottom on the figure below). The classifier uses the data to learn from
assigned labels and applies that learnt model to infer the most likely
label of the unlabelled data.</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-106"></span>
<p class="caption marginnote shownote">
Figure 12.4: Classification of samples (top) or features (bottom).
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-106-1.png" alt="Classification of samples (top) or features (bottom)." width="672">
</div>
<p>The next steps of the curriculum (course
<a href="https://uclouvain-cbio.github.io/WSBIM2122/%5D">WSBIM2122</a>) will build
upon the skills gained in this course to fully analyse complete
datasets from omics technologies, using state-of-the-art statistical
and machine learning methods and software. The course will be project
based: each experiment and associated technologies will be introduced,
the analysis pipeline will be explained, and the students will then
implement and present the data analysis and critical interpretation of
the results.</p>
<div id="additional-exercise" class="section level2" number="12.1">
<h2>
<span class="header-section-number">12.1</span> Additional exercise<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('additional-exercise')" onmouseout="reset_tooltip('additional-exercise-tooltip')"><span class="tooltiptext" id="additional-exercise-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>These final exercises integrate several techniques seen throughout
this course.</p>
<div id="exercise-1" class="section level3 unnumbered">
<h3>Exercise 1<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('exercise-1')" onmouseout="reset_tooltip('exercise-1-tooltip')"><span class="tooltiptext" id="exercise-1-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>The the RNA-Seq data and sample annotation in the files returned by
the <code>rWSBIM1207::kem2.tsv()</code> function. How many genes have been
assayed? Describe the experimental design at hand.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Print and describe the experimental design.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- Cell types A and B have been assay in a stimulted and non-stimulated -->
<!-- (none) conditions. There are 4 replicates in each of the 4 groups. -->
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Before performing log-transformation of the data, check if there are
any 0-expression values. Indeed, these would be converted to <code>-Inf</code>
after log-transformation. Are there any 0-values in the data. If so,
how many are there.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Add 1 to all expression values, then log-2 transform the expression
data. Visualise the distributions of the expression values in each
samples before and after transformation.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Continue with log-transformed data, without preforming any normalisation.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Compute, for each gene, the number of 0-expression values and
visualise these as a table showing the number of genes with 0, 1, 2,
… zero values.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise and interpret the experiment design (cell type and
treatment) using a principal component analyis.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- The PCA shows that samples group by treatment along the second PC -->
<!-- (22.3% of variance explained) and that close to 50% of the variance is -->
<!-- explained by variability along all samples. -->
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Use a t test to identify the genes that are differentially expressed
between the stimulated and non-stimulated samples of cell type
A. Visualise the results on a volcano plot.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- We start by selecting the relevant samples and create a new expression -->
<!-- matrix matching cell type A.  -->
<!-- Below, we perform a t test on each of the 4774 genes. -->
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Considering that a gene is called differentially expressed if it has
an absolute log2 fold-change &gt; 1 and an adjusted p-value &lt; 0.001, how
many differentially genes are there?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise the expression of these differentially expressed genes in
all sampes using a heatmap and interpret the figure.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- We see that most genes that are differentially expressed in cell type -->
<!-- A are show a consistent pattern in cell type B, except for the botton -->
<!-- 5 genes, that show a low expression accross both conditions. -->
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Select the 3 genes with the smallest p-values and visualise their
expression in all samples (cell types A and B, both stimulated and
un-stimulated) using boxplots.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Perform a k-means clustering of all genes and samples using k = 5 and
visualise these clusters on a PCA plot. Interprete the figure.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- The clusters represent the diversity of gene expression along -->
<!-- PC1. Cluster 3, on the left of the figure, seems to be the most -->
<!-- variable and different, while to others represent a more continuous -->
<!-- gradient along of gene expression variation. -->
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the volcano plot above, colouring the genes based on their
respective clusters. Refine your interpretation of the gene-level PCA
above.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- We see that the most diverse cluster corresponds roughly to the genes -->
<!-- that show significant down regulation upon stimulation. -->
</div>
<div id="exercise-2" class="section level3 unnumbered">
<h3>Exercise 2<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('exercise-2')" onmouseout="reset_tooltip('exercise-2-tooltip')"><span class="tooltiptext" id="exercise-2-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>recapSE1</code> data available in the <code>rWSBIM1322</code> package
(version &gt;= 0.3.1) with <code>data(recapSE1)</code>. The data contains
quantitative proteomics data. You are asked to:</p>
<ul>
<li>Familiarise yourself with the experimental design.</li>
<li>Verify if the data need to be transformed and or normalised.</li>
<li>Perform an principal component analyse and interpret it.</li>
<li>Identify differentially abundant proteins for each condition
compared to the reference group <code>CTRL</code>.</li>
<li>Are there any significantly up- or down-regulated proteins shared
between all treatments?</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="exercise-3" class="section level3 unnumbered">
<h3>Exercise 3<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('exercise-3')" onmouseout="reset_tooltip('exercise-3-tooltip')"><span class="tooltiptext" id="exercise-3-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>recapSE2</code> data available in the <code>rWSBIM1322</code> package
(version &gt;= 0.3.2) with <code>data(recapSE2)</code>. The data contains
quantitative proteomics data. You are asked to:</p>
<ul>
<li>Familiarise yourself with the experimental design.</li>
<li>Verify if the data need to be transformed and or normalised.</li>
<li>Perform an principal component analyse and interpret it.</li>
<li>Identify differentially abundant proteins for each condition
compared to the reference group <code>CTRL</code>.</li>
<li>Repeat the analyses above after removing the outlier. What is the
effect of that outlier of the differential analysis.</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div>
</div>
<p style="text-align: center;">
<a href="sec-biovis.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-si.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2024-12-09
 using 
R version 4.4.1 (2024-06-14)
</p>
</div>
</div>



</body>
</html>
