<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 3 High-level data structures | Bioinformatics" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2024-12-09" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 3 High-level data structures | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a href="sec-refresher.html" id="toc-sec-refresher"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html" id="toc-sec-vis"><span class="toc-section-number">2</span> Data visualisation</a>
<a id="active-page" href="sec-obj.html" id="toc-sec-obj"><span class="toc-section-number">3</span> High-level data structures</a><ul class="toc-sections">
<li class="toc"><a href="#introduction-1"> Introduction</a></li>
<li class="toc"><a href="#why-objects"> Why objects?</a></li>
<li class="toc"><a href="#sec-lm"> Examples in base R</a></li>
<li class="toc"><a href="#examples-from-the-bioconductor-project"> Examples from the Bioconductor project</a></li>
<li class="toc"><a href="#additional-exercises-2"> Additional exercises</a></li>
</ul>
<a href="sec-biostrings.html" id="toc-sec-biostrings"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a>
<a href="sec-norm.html" id="toc-sec-norm"><span class="toc-section-number">5</span> Data normalisation: centring, scaling, quantile normalisation</a>
<a href="sec-mlintro.html" id="toc-sec-mlintro"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a href="sec-testing.html" id="toc-sec-testing"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-dimred.html" id="toc-sec-dimred"><span class="toc-section-number">8</span> Unsupervised learning: dimensionality reduction</a>
<a href="sec-ul.html" id="toc-sec-ul"><span class="toc-section-number">9</span> Unsupervised learning: clustering</a>
<a href="sec-sl.html" id="toc-sec-sl"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html" id="toc-sec-biovis"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a href="sec-ccl.html" id="toc-sec-ccl"><span class="toc-section-number">12</span> Conclusions</a>
<a href="sec-si.html" id="toc-sec-si"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-obj" class="section level1" number="3">
<h1>
<span class="header-section-number">Chapter 3</span> High-level data structures</h1>
<p><strong>Learning objectives</strong></p>
<ul>
<li>Understand the difference between basic data structures and
high-level objects.</li>
<li>Understand the concept of high-level object.</li>
<li>Know how to learn about objects and access data from them.</li>
<li>Learn about how to import Bioconductor objects widely used in omics data
analysis.</li>
<li>Become familiar with how objects are implemented.</li>
</ul>
<div id="introduction-1" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Introduction<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('introduction-1')" onmouseout="reset_tooltip('introduction-1-tooltip')"><span class="tooltiptext" id="introduction-1-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The term <em>object</em> stems from the widely used object-oriented
programming (OOP) paradigm. OOP is however outside of scope of this
course (see part III of <span class="citation">Wickham (<label for="tufte-mn-11" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-11" class="margin-toggle">2014<span class="marginnote">Wickham, Hadley. 2014. <em>Advanced <span>R</span></em>. Chapman &amp; Hall/CRC the r Series. Taylor &amp; Francis.</span>)</span>, which is also <a href="https://adv-r.hadley.nz/">freely available
online</a>). In the <em>R refresher</em> chapter
<a href="sec-refresher.html#sec-refresher">1</a>, we have seen the base types of R, which we have
efficiently used in many situations. Sometimes, however, these aren’t
enough.</p>
<p>We have also use object, for example in chapter <a href="sec-vis.html#sec-vis">2</a>, when
generating ggplot visualisation. We have seen that it is possible to
store the output of <code>ggplot2</code> in a variable, that contain all the data
and information necessary to produce the figure.</p>
</div>
<div id="why-objects" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Why objects?<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('why-objects')" onmouseout="reset_tooltip('why-objects-tooltip')"><span class="tooltiptext" id="why-objects-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<ul>
<li><p>Combine many parts of a complex data, for example store data (and
possibly different forms thereof (see chapter <a href="sec-norm.html#sec-norm">5</a>) and
metadata together.</p></li>
<li><p>Objects are <strong>self describing</strong>: an object to store high-throughput
sequencing (HTS) data contains high-throughput sequencing
data. Replace high-throughput sequencing data by DNA sequences (see
chapter <a href="sec-biostrings.html#sec-biostrings">4</a>), proteomics, … and it will still
hold.</p></li>
<li><p>Using object enables <strong>consistency</strong><label for="tufte-sn-3" class="margin-toggle sidenote-number">3</label><input type="checkbox" id="tufte-sn-3" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">3</span> This consistency comes however at the cost of flexibility,
hence the importance of well designed objects, that fit several
use cases for the type of data at hand.</span>: whoever prepares the
HTS data, it will always have the same structure.</p></li>
</ul>
<ul>
<li>If all information is stored in a single object<label for="tufte-sn-4" class="margin-toggle sidenote-number">4</label><input type="checkbox" id="tufte-sn-4" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">4</span> This idea is consistent with the notion that you would like
to place all the data relevant to an expriment into a single
folder.</span>, it is
easier to save it, to share it with others, and to use it as an
input to a function.</li>
</ul>
<p>The above properties lead to greater <strong>robustness</strong> and
<strong>interoperability</strong> when working with objects modelling complex data.</p>
<ul>
<li><p>Common behaviour (methods) can be specialised to match the
particular data type (this is called <strong>polymorphism</strong>): for example
plotting (a vector, a matrix, NGS data), summary, subsetting (<code>[</code>
and <code>[[</code>), …</p></li>
<li><p>Another import benefit of using objects to model complex data is
<strong>abstraction</strong>: it is possible to hide all the (unnecessary for its
usage) complexity of a particular type of data and present/give
access to its most imporant parts in a user-friendly <strong>interface</strong>
and efficient way.</p></li>
<li><p>Relations between object can be recorded (this is called
<strong>inheritance</strong>).</p></li>
</ul>
<p>To avoid confusion, it is useful to define and separate objects and
classes. A <strong>class</strong> describes (models) the general properties for
groups of objects. For example, we could define a class to describe
people in this course as follows:</p>
<pre><code>Class person:
- Name: a character
- Surname: a character
- Noma: an optional numeric
- Role: a character (student, teaching assistant, instructor)</code></pre>
<p>An <strong>object</strong> is an instantiation of a class with specific values in
its fields:</p>
<pre><code>Object of class person:
- Name: Laurent
- Surname: Gatto
- Noma: NA
- Role: instructor</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Write down two objects that would describe a teaching assistant and a
student.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-7" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-7', 'sol-start-7')"></span>
</p>
<div id="sol-body-7" class="solution-body" style="display: none;">
<pre><code>Object of class person:
- Name: Axelle
- Surname: Loriot
- Noma: NA
- Role: teaching assistant

Object of class person:
- Name: Alejandra
- Surname: Sanchez-Cortez
- Noma: 123456
- Role: student</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>We will focus on objects and their usage here.</p>
</div>
<div id="sec-lm" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Examples in base R<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('sec-lm')" onmouseout="reset_tooltip('sec-lm-tooltip')"><span class="tooltiptext" id="sec-lm-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Let’s start by generating a testing dataset composed of two correlated
variables, <code>x</code> and <code>y</code>:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="sec-obj.html#cb57-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb57-2"><a href="sec-obj.html#cb57-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">100</span>)</span>
<span id="cb57-3"><a href="sec-obj.html#cb57-3" tabindex="-1"></a>y <span class="ot">&lt;-</span> x <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="dv">100</span>)</span>
<span id="cb57-4"><a href="sec-obj.html#cb57-4" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(x, y)</span>
<span id="cb57-5"><a href="sec-obj.html#cb57-5" tabindex="-1"></a><span class="fu">head</span>(df)</span></code></pre></div>
<pre><code>##            x          y
## 1 -0.6264538 -1.2468205
## 2  0.1836433  0.2257592
## 3 -0.8356286 -1.7465503
## 4  1.5952808  1.7533096
## 5  0.3295078 -0.3250769
## 6 -0.8204684  0.9468189</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="sec-obj.html#cb59-1" tabindex="-1"></a><span class="fu">plot</span>(y <span class="sc">~</span> x, df)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:df"></span>
<p class="caption marginnote shownote">
Figure 3.1: The <code>x</code> and <code>y</code> variables display a linear relationship.
</p>
<img src="WSBIM1322_files/figure-html/df-1.png" alt="The `x` and `y` variables display a linear relationship." width="672">
</div>
<p>We are now going to model the linear relation between <code>x</code> and <code>y</code>
using the <code>lm</code> (linear model) function. The resulting variable <code>fit</code>
is the result of computing the linear regression
<span class="math inline">\(y = \beta_{0} + \beta_{1} x + \epsilon\)</span>.
When printing the object, we see the call we performed, the
intercept (<span class="math inline">\(\beta_0\)</span>) and the slope (<span class="math inline">\(\beta_1\)</span>) of the regression.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="sec-obj.html#cb60-1" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> x, df)</span>
<span id="cb60-2"><a href="sec-obj.html#cb60-2" tabindex="-1"></a>fit</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = y ~ x, data = df)
## 
## Coefficients:
## (Intercept)            x  
##    -0.03769      0.99894</code></pre>
<p>We can plot the linear regression line with <code>abline</code> and passing the
<code>fit</code> variable.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="sec-obj.html#cb62-1" tabindex="-1"></a><span class="fu">plot</span>(y <span class="sc">~</span> x, df)</span>
<span id="cb62-2"><a href="sec-obj.html#cb62-2" tabindex="-1"></a><span class="fu">abline</span>(fit, <span class="at">col =</span> <span class="st">"red"</span>)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:plotlm"></span>
<p class="caption marginnote shownote">
Figure 3.2: The regression line modelling the linear regression between the two <code>x</code> and <code>y</code> variables.
</p>
<img src="WSBIM1322_files/figure-html/plotlm-1.png" alt="The regression line modelling the linear regression between the two `x` and `y` variables." width="672">
</div>
<p>The <code>fit</code> variable is an object of class <code>lm</code>, which we can check with
the <code>class</code> function.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="sec-obj.html#cb63-1" tabindex="-1"></a><span class="fu">class</span>(fit)</span></code></pre></div>
<pre><code>## [1] "lm"</code></pre>
<p>It contains much more than simply the intercept and the slope of the
regression line.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Find out what an object of class <code>lm</code> contains by exploring the
content of the <code>fit</code> object (try the <code>names</code> and <code>str</code> functions) and
reading the <code>lm</code> manual page.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Finally, to demonstrate <strong>polymorphism</strong>, let’s try to plot the <code>fit</code>
variable.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="sec-obj.html#cb65-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb65-2"><a href="sec-obj.html#cb65-2" tabindex="-1"></a><span class="fu">plot</span>(fit)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:lmplot"></span>
<p class="caption marginnote shownote">
Figure 3.3: Plotting an instance of class <code>lm</code> produces a set of diagnostic plots, that inform us about the quality of the regression.
</p>
<img src="WSBIM1322_files/figure-html/lmplot-1.png" alt="Plotting an instance of class `lm` produces a set of diagnostic plots, that inform us about the quality of the regression." width="960">
</div>
</div>
<div id="examples-from-the-bioconductor-project" class="section level2" number="3.4">
<h2>
<span class="header-section-number">3.4</span> Examples from the Bioconductor project<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('examples-from-the-bioconductor-project')" onmouseout="reset_tooltip('examples-from-the-bioconductor-project-tooltip')"><span class="tooltiptext" id="examples-from-the-bioconductor-project-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>We have already briefly seen the <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-bioinfo.html#sec-bioconductor">Bioconductor
project</a>. It
initiated by Robert Gentleman (<span class="citation">Gentleman et al. (<label for="tufte-mn-12" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-12" class="margin-toggle">2004<span class="marginnote">Gentleman, Robert C., Vincent J. Carey, Douglas M. Bates, Ben Bolstad, Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004. <span>“Bioconductor: Open Software Development for Computational Biology and Bioinformatics.”</span> <em>Genome Biol</em> 5 (10): –80. <a href="https://doi.org/10.1186/gb-2004-5-10-r80">https://doi.org/10.1186/gb-2004-5-10-r80</a>.</span>)</span>; <span class="citation">Huber et al. (<label for="tufte-mn-13" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-13" class="margin-toggle">2015<span class="marginnote">Huber, W, V J Carey, R Gentleman, S Anders, M Carlson, B S Carvalho, H C Bravo, et al. 2015. <span>“Orchestrating High-Throughput Genomic Analysis with <span>Bioconductor</span>.”</span> <em>Nat Methods</em> 12 (2): 115–21. <a href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>.</span>)</span>), one of
the two creators of the R language, and centrally offers dedicated R
packages for bioinformatics.</p>
<blockquote>
<p>Bioconductor provides tools for the analysis and comprehension of
high-throughput genomic data. Bioconductor uses the R statistical
programming language, and is open source and open development. It
has two releases each year, and an active user community.</p>
</blockquote>
<p>This <a href="https://www.youtube.com/watch?v=nzY7bPQOXUs">video</a> provides a
great overview of the project at large.</p>
<p>Bioconductor provides a rich set of <a href="https://bioconductor.org/developers/how-to/commonMethodsAndClasses/">classes for omics
data</a>,
summarised below.</p>
<div id="common-bioconductor-classes" class="section level3" number="3.4.1">
<h3>
<span class="header-section-number">3.4.1</span> Common Bioconductor classes<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('common-bioconductor-classes')" onmouseout="reset_tooltip('common-bioconductor-classes-tooltip')"><span class="tooltiptext" id="common-bioconductor-classes-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<ul>
<li>Rectangular feature x sample data – <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment">SummarizedExperiment</a></em><code>::SummarizedExperiment()</code> (RNAseq count matrix, microarray, …)</li>
<li>Genomic coordinates – <em><a href="https://bioconductor.org/packages/3.20/GenomicRanges">GenomicRanges</a></em><code>::GRanges()</code> (1-based, closed interval)</li>
<li>DNA / RNA / AA sequences – <em><a href="https://bioconductor.org/packages/3.20/Biostrings">Biostrings</a></em><code>::*StringSet()</code>
</li>
<li>Gene sets – <em><a href="https://bioconductor.org/packages/3.20/GSEABase">GSEABase</a></em><code>::GeneSet()</code> <em><a href="https://bioconductor.org/packages/3.20/GSEABase">GSEABase</a></em><code>::GeneSetCollection()</code>
</li>
<li>Multi-omics data – <em><a href="https://bioconductor.org/packages/3.20/MultiAssayExperiment">MultiAssayExperiment</a></em><code>::MultiAssayExperiment()</code>
</li>
<li>Single cell data – <em><a href="https://bioconductor.org/packages/3.20/SingleCellExperiment">SingleCellExperiment</a></em><code>::SingleCellExperiment()</code>
</li>
<li>Mass spectrometry data – <em><a href="https://bioconductor.org/packages/3.20/Spectra">Spectra</a></em><code>::Spectra()</code>
</li>
<li>Quantitative proteomics data – <em><a href="https://bioconductor.org/packages/3.20/QFeatures">QFeatures</a></em><code>::QFeatures()</code>
</li>
</ul>
</div>
<div id="the-summarizedexperiment-class" class="section level3" number="3.4.2">
<h3>
<span class="header-section-number">3.4.2</span> The <code>SummarizedExperiment</code> class<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('the-summarizedexperiment-class')" onmouseout="reset_tooltip('the-summarizedexperiment-class-tooltip')"><span class="tooltiptext" id="the-summarizedexperiment-class-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Let’s start by installing<label for="tufte-sn-5" class="margin-toggle sidenote-number">5</label><input type="checkbox" id="tufte-sn-5" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">5</span> Before installing the package, we first check if they are
available with <code>require</code>. If they are missing, the call to
<code>require</code> will return <code>FALSE</code>, otherwise it returns <code>TRUE</code> and
loads the packages. Given that we negate the return value of
<code>require</code> using <code>!</code>, the packages will be installed only when
missing.</span> the<em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment">SummarizedExperiment</a></em>
package, that implements the class, and the <em><a href="https://bioconductor.org/packages/3.20/airway">airway</a></em>
data package, that provides an RNA-Seq example data.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="sec-obj.html#cb66-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"SummarizedExperiment"</span>))</span>
<span id="cb66-2"><a href="sec-obj.html#cb66-2" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"SummarizedExperiment"</span>)</span>
<span id="cb66-3"><a href="sec-obj.html#cb66-3" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"airway"</span>))</span>
<span id="cb66-4"><a href="sec-obj.html#cb66-4" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"airway"</span>)</span></code></pre></div>
<p>If you know you have the packages, it is more straightforward to
directly load the packages.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="sec-obj.html#cb67-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"SummarizedExperiment"</span>)</span>
<span id="cb67-2"><a href="sec-obj.html#cb67-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"airway"</span>)</span></code></pre></div>
<p>The <code>airway</code> package provides a data for an RNA-Seq experiment on four
human airway smooth muscle cell lines treated with dexamethason
published by <span class="citation">Himes et al. (<label for="tufte-mn-14" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-14" class="margin-toggle">2014<span class="marginnote">Himes, B E, X Jiang, P Wagner, R Hu, Q Wang, B Klanderman, R M Whitaker, et al. 2014. <span>“RNA-Seq Transcriptome Profiling Identifies CRISPLD2 as a Glucocorticoid Responsive Gene That Modulates Cytokine Function in Airway Smooth Muscle Cells.”</span> <em>PLoS One</em> 9 (6): e99625. <a href="https://doi.org/10.1371/journal.pone.0099625">https://doi.org/10.1371/journal.pone.0099625</a>.</span>)</span>. It contains the RNA expression data for
64102 transcripts and 8 samples. These 8 samples correspond to treated
and untreated condition in 4 different cell lines.</p>
<p>Objects of the class <code>SummarizedExperiment</code> contain one or more assays
(i.e. sets if quantitative expresson data), each represented by a
matrix-like object (typically) of numeric mode. The rows represent
features such as genes, transcripts, proteins or more generally
genomic ranges and the columns represent samples. The figure below
represents its anatomy in greater details.</p>
<div class="figure">
<span style="display:block;" id="fig:sefig"></span>
<p class="caption marginnote shownote">
Figure 3.4: Schematic representation of the anatomy if a <code>SummarizedExperiment</code> object. (Figure taken from the <code>SummarizedExperiment</code> package vignette.)
</p>
<img src="figs/SE.svg" alt="Schematic representation of the anatomy if a `SummarizedExperiment` object. (Figure taken from the `SummarizedExperiment` package vignette.)">
</div>
<ul>
<li>The sample (columns) metadata can be accessed with the <code>colData()</code>
function.</li>
<li>The features (rows) metadata can be accessed with the <code>rowData()</code>
column.</li>
<li>If the features represent ranges along genomic coordinates, these
can be accessed with <code>rowRanges()</code>
</li>
<li>Additional metadata describing the overall experiment can be
accessed with <code>metadata()</code>.</li>
<li>The quantiative data can be accessed with <code>assay()</code>.</li>
<li>
<code>assays()</code> returns a list of matrix-like assays.</li>
</ul>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>airway</code> data and display it in your R console. Make sure you
understand all parts in the short text summary that is displayed. What
class is <code>airway</code> of.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Extract the quantitative assay data, the rows and columns metadata,
the row ranges and the experiment metadata. What are the classes of
these individual parts of the <code>SummarizedExperiment</code> object.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-8" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-8', 'sol-start-8')"></span>
</p>
<div id="sol-body-8" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="sec-obj.html#cb68-1" tabindex="-1"></a><span class="fu">data</span>(airway)</span>
<span id="cb68-2"><a href="sec-obj.html#cb68-2" tabindex="-1"></a>airway</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 63677 8 
## metadata(1): ''
## assays(1): counts
## rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492
##   ENSG00000273493
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="sec-obj.html#cb70-1" tabindex="-1"></a><span class="fu">class</span>(airway)</span></code></pre></div>
<pre><code>## [1] "RangedSummarizedExperiment"
## attr(,"package")
## [1] "SummarizedExperiment"</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="sec-obj.html#cb72-1" tabindex="-1"></a><span class="fu">assay</span>(airway)</span></code></pre></div>
<pre><code>##                 SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516
## ENSG00000000003        679        448        873        408       1138
## ENSG00000000005          0          0          0          0          0
## ENSG00000000419        467        515        621        365        587
## ENSG00000000457        260        211        263        164        245
## ENSG00000000460         60         55         40         35         78
## ENSG00000000938          0          0          2          0          1
## ENSG00000000971       3251       3679       6177       4252       6721
## ENSG00000001036       1433       1062       1733        881       1424
## ENSG00000001084        519        380        595        493        820
## ENSG00000001167        394        236        464        175        658
## ENSG00000001460        172        168        264        118        241
## ENSG00000001461       2112       1867       5137       2657       2735
##                 SRR1039517 SRR1039520 SRR1039521
## ENSG00000000003       1047        770        572
## ENSG00000000005          0          0          0
## ENSG00000000419        799        417        508
## ENSG00000000457        331        233        229
## ENSG00000000460         63         76         60
## ENSG00000000938          0          0          0
## ENSG00000000971      11027       5176       7995
## ENSG00000001036       1439       1359       1109
## ENSG00000001084        714        696        704
## ENSG00000001167        584        360        269
## ENSG00000001460        210        155        177
## ENSG00000001461       2751       2467       2905
##  [ reached getOption("max.print") -- omitted 63665 rows ]</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="sec-obj.html#cb74-1" tabindex="-1"></a><span class="fu">class</span>(<span class="fu">assay</span>(airway))</span></code></pre></div>
<pre><code>## [1] "matrix" "array"</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="sec-obj.html#cb76-1" tabindex="-1"></a><span class="fu">colData</span>(airway)</span></code></pre></div>
<pre><code>## DataFrame with 8 rows and 9 columns
##            SampleName     cell      dex    albut        Run avgLength
##              &lt;factor&gt; &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;   &lt;factor&gt; &lt;integer&gt;
## SRR1039508 GSM1275862  N61311     untrt    untrt SRR1039508       126
## SRR1039509 GSM1275863  N61311     trt      untrt SRR1039509       126
## SRR1039512 GSM1275866  N052611    untrt    untrt SRR1039512       126
## SRR1039513 GSM1275867  N052611    trt      untrt SRR1039513        87
## SRR1039516 GSM1275870  N080611    untrt    untrt SRR1039516       120
## SRR1039517 GSM1275871  N080611    trt      untrt SRR1039517       126
## SRR1039520 GSM1275874  N061011    untrt    untrt SRR1039520       101
## SRR1039521 GSM1275875  N061011    trt      untrt SRR1039521        98
##            Experiment    Sample    BioSample
##              &lt;factor&gt;  &lt;factor&gt;     &lt;factor&gt;
## SRR1039508  SRX384345 SRS508568 SAMN02422669
## SRR1039509  SRX384346 SRS508567 SAMN02422675
## SRR1039512  SRX384349 SRS508571 SAMN02422678
## SRR1039513  SRX384350 SRS508572 SAMN02422670
## SRR1039516  SRX384353 SRS508575 SAMN02422682
## SRR1039517  SRX384354 SRS508576 SAMN02422673
## SRR1039520  SRX384357 SRS508579 SAMN02422683
## SRR1039521  SRX384358 SRS508580 SAMN02422677</code></pre>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="sec-obj.html#cb78-1" tabindex="-1"></a><span class="fu">class</span>(<span class="fu">colData</span>(airway))</span></code></pre></div>
<pre><code>## [1] "DFrame"
## attr(,"package")
## [1] "S4Vectors"</code></pre>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="sec-obj.html#cb80-1" tabindex="-1"></a><span class="fu">rowData</span>(airway)</span></code></pre></div>
<pre><code>## DataFrame with 63677 rows and 10 columns
##                         gene_id     gene_name  entrezid   gene_biotype
##                     &lt;character&gt;   &lt;character&gt; &lt;integer&gt;    &lt;character&gt;
## ENSG00000000003 ENSG00000000003        TSPAN6        NA protein_coding
## ENSG00000000005 ENSG00000000005          TNMD        NA protein_coding
## ENSG00000000419 ENSG00000000419          DPM1        NA protein_coding
## ENSG00000000457 ENSG00000000457         SCYL3        NA protein_coding
## ENSG00000000460 ENSG00000000460      C1orf112        NA protein_coding
## ...                         ...           ...       ...            ...
## ENSG00000273489 ENSG00000273489 RP11-180C16.1        NA      antisense
## ENSG00000273490 ENSG00000273490        TSEN34        NA protein_coding
## ENSG00000273491 ENSG00000273491  RP11-138A9.2        NA        lincRNA
##                 gene_seq_start gene_seq_end              seq_name seq_strand
##                      &lt;integer&gt;    &lt;integer&gt;           &lt;character&gt;  &lt;integer&gt;
## ENSG00000000003       99883667     99894988                     X         -1
## ENSG00000000005       99839799     99854882                     X          1
## ENSG00000000419       49551404     49575092                    20         -1
## ENSG00000000457      169818772    169863408                     1         -1
## ENSG00000000460      169631245    169823221                     1          1
## ...                        ...          ...                   ...        ...
## ENSG00000273489      131178723    131182453                     7         -1
## ENSG00000273490       54693789     54697585 HSCHR19LRC_LRC_J_CTG1          1
## ENSG00000273491      130600118    130603315          HG1308_PATCH          1
##                 seq_coord_system        symbol
##                        &lt;integer&gt;   &lt;character&gt;
## ENSG00000000003               NA        TSPAN6
## ENSG00000000005               NA          TNMD
## ENSG00000000419               NA          DPM1
## ENSG00000000457               NA         SCYL3
## ENSG00000000460               NA      C1orf112
## ...                          ...           ...
## ENSG00000273489               NA RP11-180C16.1
## ENSG00000273490               NA        TSEN34
## ENSG00000273491               NA  RP11-138A9.2
##  [ reached getOption("max.print") -- omitted 2 rows ]</code></pre>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="sec-obj.html#cb82-1" tabindex="-1"></a><span class="fu">class</span>(<span class="fu">rowData</span>(airway))</span></code></pre></div>
<pre><code>## [1] "DFrame"
## attr(,"package")
## [1] "S4Vectors"</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="sec-obj.html#cb84-1" tabindex="-1"></a><span class="fu">rowRanges</span>(airway)</span></code></pre></div>
<pre><code>## GRangesList object of length 63677:
## $ENSG00000000003
## GRanges object with 17 ranges and 2 metadata columns:
##        seqnames            ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]        X 99883667-99884983      - |    667145 ENSE00001459322
##    [2]        X 99885756-99885863      - |    667146 ENSE00000868868
##    [3]        X 99887482-99887565      - |    667147 ENSE00000401072
##    [4]        X 99887538-99887565      - |    667148 ENSE00001849132
##    [5]        X 99888402-99888536      - |    667149 ENSE00003554016
##    ...      ...               ...    ... .       ...             ...
##   [13]        X 99890555-99890743      - |    667156 ENSE00003512331
##   [14]        X 99891188-99891686      - |    667158 ENSE00001886883
##   [15]        X 99891605-99891803      - |    667159 ENSE00001855382
##   [16]        X 99891790-99892101      - |    667160 ENSE00001863395
##   [17]        X 99894942-99894988      - |    667161 ENSE00001828996
##   -------
##   seqinfo: 722 sequences (1 circular) from an unspecified genome
## 
## ...
## &lt;63676 more elements&gt;</code></pre>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="sec-obj.html#cb86-1" tabindex="-1"></a><span class="fu">class</span>(<span class="fu">rowRanges</span>(airway))</span></code></pre></div>
<pre><code>## [1] "CompressedGRangesList"
## attr(,"package")
## [1] "GenomicRanges"</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The row and column metadata are stored as a special <code>data.frame</code>
implementation<label for="tufte-sn-6" class="margin-toggle sidenote-number">6</label><input type="checkbox" id="tufte-sn-6" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">6</span> Like the <code>tibble</code>, that is an implememtation specific to the
<code>tidyverse</code> packages.</span>, that has been developed in the Bioconductor
project. It behaves like like a classical <code>data.frame</code>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>What biological sample names in the <code>airway</code> dataset were treated with
dexamethason? Which ones weren’t?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-9" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-9', 'sol-start-9')"></span>
</p>
<div id="sol-body-9" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="sec-obj.html#cb88-1" tabindex="-1"></a>dex_treated <span class="ot">&lt;-</span> airway<span class="sc">$</span>dex <span class="sc">==</span> <span class="st">"trt"</span></span>
<span id="cb88-2"><a href="sec-obj.html#cb88-2" tabindex="-1"></a><span class="do">## shorthand for colData(airway)$dex == "trt"</span></span>
<span id="cb88-3"><a href="sec-obj.html#cb88-3" tabindex="-1"></a><span class="do">## Treated samples</span></span>
<span id="cb88-4"><a href="sec-obj.html#cb88-4" tabindex="-1"></a><span class="fu">colData</span>(airway)[dex_treated, <span class="st">"BioSample"</span>]</span></code></pre></div>
<pre><code>## [1] SAMN02422675 SAMN02422670 SAMN02422673 SAMN02422677
## 8 Levels: SAMN02422669 SAMN02422670 SAMN02422673 SAMN02422675 ... SAMN02422683</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="sec-obj.html#cb90-1" tabindex="-1"></a><span class="do">## Control samples</span></span>
<span id="cb90-2"><a href="sec-obj.html#cb90-2" tabindex="-1"></a><span class="fu">colData</span>(airway)[<span class="sc">!</span>dex_treated, <span class="st">"BioSample"</span>]</span></code></pre></div>
<pre><code>## [1] SAMN02422669 SAMN02422678 SAMN02422682 SAMN02422683
## 8 Levels: SAMN02422669 SAMN02422670 SAMN02422673 SAMN02422675 ... SAMN02422683</code></pre>
<p>It is also possible to use the <code>tidyverse</code> for this by first coercing
the <code>DataFrame</code> to a <code>tibble</code>.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="sec-obj.html#cb92-1" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">colData</span>(airway)) <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="sec-obj.html#cb92-2" tabindex="-1"></a>    <span class="fu">filter</span>(dex <span class="sc">==</span> <span class="st">"trt"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb92-3"><a href="sec-obj.html#cb92-3" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(BioSample)</span></code></pre></div>
<pre><code>## # A tibble: 4 × 1
##   BioSample   
##   &lt;fct&gt;       
## 1 SAMN02422675
## 2 SAMN02422670
## 3 SAMN02422673
## 4 SAMN02422677</code></pre>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="sec-obj.html#cb94-1" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">colData</span>(airway)) <span class="sc">%&gt;%</span></span>
<span id="cb94-2"><a href="sec-obj.html#cb94-2" tabindex="-1"></a>    <span class="fu">filter</span>(dex <span class="sc">!=</span> <span class="st">"trt"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb94-3"><a href="sec-obj.html#cb94-3" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(BioSample)</span></code></pre></div>
<pre><code>## # A tibble: 4 × 1
##   BioSample   
##   &lt;fct&gt;       
## 1 SAMN02422669
## 2 SAMN02422678
## 3 SAMN02422682
## 4 SAMN02422683</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>What are the dimensions of the assay data, and the row and column
metadata? What can you way about these?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-10" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-10', 'sol-start-10')"></span>
</p>
<div id="sol-body-10" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="sec-obj.html#cb96-1" tabindex="-1"></a><span class="fu">dim</span>(airway)</span></code></pre></div>
<pre><code>## [1] 63677     8</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="sec-obj.html#cb98-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">colData</span>(airway))</span></code></pre></div>
<pre><code>## [1] 8 9</code></pre>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="sec-obj.html#cb100-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">rowData</span>(airway))</span></code></pre></div>
<pre><code>## [1] 63677    10</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="subsetting-summarizedexperiment-objects" class="section level3" number="3.4.3">
<h3>
<span class="header-section-number">3.4.3</span> Subsetting <code>SummarizedExperiment</code> objects<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('subsetting-summarizedexperiment-objects')" onmouseout="reset_tooltip('subsetting-summarizedexperiment-objects-tooltip')"><span class="tooltiptext" id="subsetting-summarizedexperiment-objects-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>As mentioned in the introduction, it is possible to redefine common
functions to have a special behaviour with objects. This is the case
for the <code>[</code> operator, that will automatically subset all parts of the
object. Below, we create a new instance of class
<code>SummarizedExperiment</code> that contains only the 5 first features for the
3 first samples.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="sec-obj.html#cb102-1" tabindex="-1"></a>se <span class="ot">&lt;-</span> airway[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb102-2"><a href="sec-obj.html#cb102-2" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 5 3 
## metadata(1): ''
## assays(1): counts
## rownames(5): ENSG00000000003 ENSG00000000005 ENSG00000000419
##   ENSG00000000457 ENSG00000000460
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(3): SRR1039508 SRR1039509 SRR1039512
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Verfiy that the assay data, the column and row metadata and the row
ranges have been preserved.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Subsetting can be performed with numeric vectors (as shown above), as
well as character of logical vectors, just like with dataframes or
matrices.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Create a new object that contains the data for features (rows)
<code>"ENSG00000166963"</code>, <code>"ENSG00000267970"</code> and <code>"ENSG00000254663"</code> and
samples (columns) <code>"SRR1039517"</code> and <code>"SRR1039521"</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-11" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-11', 'sol-start-11')"></span>
</p>
<div id="sol-body-11" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="sec-obj.html#cb104-1" tabindex="-1"></a>airway[<span class="fu">c</span>(<span class="st">"ENSG00000166963"</span>, <span class="st">"ENSG00000267970"</span>, <span class="st">"ENSG00000254663"</span>),</span>
<span id="cb104-2"><a href="sec-obj.html#cb104-2" tabindex="-1"></a>       <span class="fu">c</span>(<span class="st">"SRR1039517"</span>, <span class="st">"SRR1039521"</span>)]</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 3 2 
## metadata(1): ''
## assays(1): counts
## rownames(3): ENSG00000166963 ENSG00000267970 ENSG00000254663
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(2): SRR1039517 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The row ranges are defined as a list of genomic ranges (<code>GRanges</code>
objects), one genomic range per row in the
<code>SummarizedExperiment</code>. This <code>GRangesList</code> object behaves like a list,
albeit a list that specifically contains <code>GRanges</code> objects.</p>
<p>If we look at the first feature, corresponding to gene
<code>ENSG00000000003</code>, we see that is is composed of 17 exons
(<code>ENSE00001459322</code> to <code>ENSE00001828996</code>) and their detailed positions
on chromosome X.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="sec-obj.html#cb106-1" tabindex="-1"></a><span class="fu">rowRanges</span>(airway)[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## GRanges object with 17 ranges and 2 metadata columns:
##        seqnames            ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]        X 99883667-99884983      - |    667145 ENSE00001459322
##    [2]        X 99885756-99885863      - |    667146 ENSE00000868868
##    [3]        X 99887482-99887565      - |    667147 ENSE00000401072
##    [4]        X 99887538-99887565      - |    667148 ENSE00001849132
##    [5]        X 99888402-99888536      - |    667149 ENSE00003554016
##    ...      ...               ...    ... .       ...             ...
##   [13]        X 99890555-99890743      - |    667156 ENSE00003512331
##   [14]        X 99891188-99891686      - |    667158 ENSE00001886883
##   [15]        X 99891605-99891803      - |    667159 ENSE00001855382
##   [16]        X 99891790-99892101      - |    667160 ENSE00001863395
##   [17]        X 99894942-99894988      - |    667161 ENSE00001828996
##   -------
##   seqinfo: 722 sequences (1 circular) from an unspecified genome</code></pre>
<p>The position of genes can also be used to subset a
<code>SummarizedExperiment</code>. In the example below, we select all features
that overlap with positions 100000 to 1100000 on chromosome 1. To do
this we first create a new <code>GRanges</code> object that matches the region of
interest:</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="sec-obj.html#cb108-1" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames=</span><span class="st">"1"</span>, <span class="at">ranges=</span><span class="dv">100000</span><span class="sc">:</span><span class="dv">1100000</span>)</span>
<span id="cb108-2"><a href="sec-obj.html#cb108-2" tabindex="-1"></a>roi</span></code></pre></div>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames         ranges strand
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]        1 100000-1100000      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>We can now use <code>roi</code> to extract the feature of interest with the
<code>subsetByOverlaps</code> function, which returns a new
<code>SummarizedExperiment</code> object that contains 74 features.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="sec-obj.html#cb110-1" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">subsetByOverlaps</span>(airway, roi)</span>
<span id="cb110-2"><a href="sec-obj.html#cb110-2" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 74 8 
## metadata(1): ''
## assays(1): counts
## rownames(74): ENSG00000131591 ENSG00000177757 ... ENSG00000272512
##   ENSG00000273443
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>How many genes are present on chromosome X between position 100000 et 1000000.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-12" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-12', 'sol-start-12')"></span>
</p>
<div id="sol-body-12" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="sec-obj.html#cb112-1" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">"X"</span>, <span class="at">ranges =</span> <span class="fl">1e5</span><span class="sc">:</span><span class="fl">1e6</span>)</span>
<span id="cb112-2"><a href="sec-obj.html#cb112-2" tabindex="-1"></a><span class="fu">nrow</span>(<span class="fu">subsetByOverlaps</span>(airway, roi))</span></code></pre></div>
<pre><code>## [1] 11</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visit the Ensembl page and look up the Ensembl gene identifier
(<code>ENSG00...</code>) of the BRCA1 gene. Verify if these match the information
in the <code>airway</code> data. If not, why could this be? How many exons are
there in the <code>airway</code> data for this gene?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-13" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-13', 'sol-start-13')"></span>
</p>
<div id="sol-body-13" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="sec-obj.html#cb114-1" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(airway[<span class="st">"ENSG00000012048"</span>, ])</span>
<span id="cb114-2"><a href="sec-obj.html#cb114-2" tabindex="-1"></a>gr</span></code></pre></div>
<pre><code>## GRangesList object of length 1:
## $ENSG00000012048
## GRanges object with 99 ranges and 2 metadata columns:
##        seqnames            ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]       17 41196312-41197819      - |    550890 ENSE00001814242
##    [2]       17 41196313-41197819      - |    550891 ENSE00001312675
##    [3]       17 41196822-41197819      - |    550892 ENSE00001831829
##    [4]       17 41197580-41197819      - |    550893 ENSE00002914501
##    [5]       17 41197646-41197819      - |    550894 ENSE00001937547
##    ...      ...               ...    ... .       ...             ...
##   [95]       17 41277294-41277376      - |    550984 ENSE00001896310
##   [96]       17 41277294-41277387      - |    550985 ENSE00001878709
##   [97]       17 41277294-41277419      - |    550986 ENSE00001869098
##   [98]       17 41277294-41277467      - |    550987 ENSE00001954850
##   [99]       17 41277294-41277468      - |    550988 ENSE00001601762
##   -------
##   seqinfo: 722 sequences (1 circular) from an unspecified genome</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="sec-obj.html#cb116-1" tabindex="-1"></a><span class="fu">length</span>(gr[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>## [1] 99</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="constructing-a-summarizedexperiment" class="section level3" number="3.4.4">
<h3>
<span class="header-section-number">3.4.4</span> Constructing a <code>SummarizedExperiment</code>
<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('constructing-a-summarizedexperiment')" onmouseout="reset_tooltip('constructing-a-summarizedexperiment-tooltip')"><span class="tooltiptext" id="constructing-a-summarizedexperiment-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Complex object such as <code>SummarizedExperiment</code> are generally
constructed by functions (called <strong>constructors</strong>) from files. It is
possible to construct a <code>SummarizedExperiment</code> by hand. To do so, we
first need to construct the different parts of the object.</p>
<p>Let’s create an expression matrix for 200 genes and 6 samples by
randomly sample values between 1 and 10000.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="sec-obj.html#cb118-1" tabindex="-1"></a>nrows <span class="ot">&lt;-</span> <span class="dv">200</span></span>
<span id="cb118-2"><a href="sec-obj.html#cb118-2" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb118-3"><a href="sec-obj.html#cb118-3" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">runif</span>(nrows <span class="sc">*</span> ncols, <span class="dv">1</span>, <span class="fl">1e4</span>), nrows)</span></code></pre></div>
<p>Below, we create a the <code>colData</code> part using the appropriate
constructor for the class of data needed:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="sec-obj.html#cb119-1" tabindex="-1"></a>cd <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="at">Treatment =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"ChIP"</span>, <span class="st">"Input"</span>), <span class="dv">3</span>),</span>
<span id="cb119-2"><a href="sec-obj.html#cb119-2" tabindex="-1"></a>                <span class="at">row.names =</span> LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span></code></pre></div>
<p>Let’s also create some ranges. Here, we will have one range per
features, as opposed to the <code>airway</code> data, that has multiple ranges
(exons) per feature (gene). The ranges below to chromosomes 1 and 2
(50 and 150 respectively), with ranges (constructure with the
<code>IRanges</code> constructor) starting between 1e5 and 1e6 with width 100,
and strand chosing randomly as the positive or negative strand.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="sec-obj.html#cb120-1" tabindex="-1"></a>rng <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="fu">c</span>(<span class="dv">50</span>, <span class="dv">150</span>)),</span>
<span id="cb120-2"><a href="sec-obj.html#cb120-2" tabindex="-1"></a>               <span class="fu">IRanges</span>(<span class="fu">floor</span>(<span class="fu">runif</span>(<span class="dv">200</span>, <span class="fl">1e5</span>, <span class="fl">1e6</span>)), <span class="at">width =</span> <span class="dv">100</span>),</span>
<span id="cb120-3"><a href="sec-obj.html#cb120-3" tabindex="-1"></a>               <span class="at">strand =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">200</span>, <span class="cn">TRUE</span>))</span></code></pre></div>
<p>We can now put the different parts together using the
<code>SummarizedExperiment</code> constructor:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="sec-obj.html#cb121-1" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> counts),</span>
<span id="cb121-2"><a href="sec-obj.html#cb121-2" tabindex="-1"></a>                           <span class="at">colData =</span> cd,</span>
<span id="cb121-3"><a href="sec-obj.html#cb121-3" tabindex="-1"></a>                           <span class="at">rowRanges =</span> rng)</span>
<span id="cb121-4"><a href="sec-obj.html#cb121-4" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 200 6 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(0):
## colnames(6): A B ... E F
## colData names(1): Treatment</code></pre>
<p>We can also create a <code>rowData</code> object and add it to our <code>se</code> object:</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="sec-obj.html#cb123-1" tabindex="-1"></a><span class="fu">rowData</span>(se) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="at">some_value =</span> <span class="fu">sample</span>(<span class="dv">200</span>))</span>
<span id="cb123-2"><a href="sec-obj.html#cb123-2" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 200 6 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(1): some_value
## colnames(6): A B ... E F
## colData names(1): Treatment</code></pre>
</div>
</div>
<div id="additional-exercises-2" class="section level2" number="3.5">
<h2>
<span class="header-section-number">3.5</span> Additional exercises<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('additional-exercises-2')" onmouseout="reset_tooltip('additional-exercises-2-tooltip')"><span class="tooltiptext" id="additional-exercises-2-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Using the function <code>kem2.tsv</code> from the <code>rWSBIM1207</code> package, import
the <em>count</em> and <em>annotation</em> data into R and use them to create an object
of class <code>SummarizedExperiment</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div>
<p style="text-align: center;">
<a href="sec-vis.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-biostrings.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2024-12-09
 using 
R version 4.4.1 (2024-06-14)
</p>
</div>
</div>



</body>
</html>
